_posts/2021-08-31-rate-limiting-with-springboot-resilience4j.md

```diff
@@ -0,0 +1,496 @@
+
+Since we don't have a reference to the `RateLimiter` instance when working with Spring Boot Resilience4j, this requires a little more work. The idea is still the same, but how we get a reference to the `RateLimiterRegistry` and then the `RateLimiter` instance itself is a bit different.
+
+First, we `@Autowire` a `RateLimiterRegistry` into the bean that invokes the remote operation:
+
+```java
+@Service
+public class RateLimitingService {
+  @Autowired
+  private FlightSearchService remoteSearchService;
+
+  @Autowired
+  private RateLimiterRegistry registry;
+
+  // other lines omitted
+}
+```
+
+Then we add a `@PostConstruct` method which sets up the `onSuccess` and  `onFailure` event handlers:
+
+```java
+@PostConstruct
+public void postConstruct() {
+  io.github.resilience4j.ratelimiter.RateLimiter.EventPublisher eventPublisher = registry
+        .rateLimiter("rateLimiterEventsExample")
+        .getEventPublisher();
+  
+  eventPublisher.onSuccess(System.out::println);
+  eventPublisher.onFailure(System.out::println);
+}
+```
+
+Here, we fetched the `RateLimiter` instance by name from the `RateLimiterRegistry` and then got the `EventPublisher` from the `RateLimiter` instance.
+
+Instead of the `@PostConstruct` method, we could have also done the same in the constructor of `RateLimitingService`.
+
+Now, the sample output shows details of the events:
+
+```shell
+RateLimiterEvent{type=SUCCESSFUL_ACQUIRE, rateLimiterName='rateLimiterEventsExample', creationTime=2021-08-29T18:52:19.229460}
+Searching for flights; current time = 18:52:19 241
+Flight search successful
+[Flight{flightNumber='XY 765', flightDate='08/15/2021', from='NYC', to='LAX'}, ... }]
+RateLimiterEvent{type=FAILED_ACQUIRE, rateLimiterName='rateLimiterEventsExample', creationTime=2021-08-29T18:52:19.329324}
+RateLimiter 'rateLimiterEventsExample' does not permit further calls
+```
+
+## Actuator Endpoints
+
+Spring Boot Resilience4j makes the details about the last 100 rate limit events available through the Actuator endpoint `/actuator/ratelimiterevents`. Apart from this, it exposes a few other endpoints:
+
+1. `/actuator/ratelimiters`
+2. `/actuator/metrics/resilience4j.ratelimiter.available.permissions`
+3. `/actuator/metrics/resilience4j.ratelimiter.waiting_threads`
+
+Let's look at the data returned by doing a `curl` to these endpoints.
+
+### Endpoint `/actuator/ratelimiters`
+
+This endpoint lists the names of all the ratelimiter instances available:
+
+```bash
+$ curl http://localhost:8080/actuator/ratelimiters
+{
+  "rateLimiters": [
+    "basicExample",
+    "changeLimitsExample",
+    "multipleRateLimiters_rpm_limiter",
+    "multipleRateLimiters_rps_limiter",
+    "rateLimiterEventsExample",
+    "retryAndRateLimitExample",
+    "timeoutExample",
+    "fallbackExample"
+  ]
+}
+```
+
+### Endpoint `/actuator/metrics/resilience4j.ratelimiter.available.permissions`
```

Comment:
@thombergs this header and the next one are too long to render properly on screen. I was not sure how to make it smaller. Maybe we can introduce them as "Endpoint 1 - ...", "Endpoint 2 - /actuator..." in the section intro and call them "Endpoint 1", "Endpoint 2" etc. here?

Reply:
Yeah, I changed that to "Permissions Endpoint" etc.